[Automation] Test cases for volume create, access and vsan

pshahzeb opened the pull request:
1. Test cases for volume create with valid/invalid volume names
2. Test cases for volume create with valid/invalid volume options
3. Test cases for volume access updates and consistently test read/write operations
4. Test cases for vsan volume

Fixes #1254 

Testing: done locally
test-all target passes
e2-test output
```
=== RUN   Test
2017/05/25 22:48:10 START: basic-test.TestVolumeIsolation
2017/05/25 22:48:10 Creating volume [basic_test_volume_1495752490] on VM [10.192.51.202]
2017/05/25 22:48:13 Checking volume [basic_test_volume_1495752490] availability from VM [10.192.51.202]
2017/05/25 22:48:13 Checking volume [basic_test_volume_1495752490] availability from VM [10.192.51.202]
2017/05/25 22:48:14 Destroying volume [basic_test_volume_1495752490]
2017/05/25 22:48:15 END: basic-test.TestVolumeIsolation
2017/05/25 22:48:15 START: basic-test.TestVolumeLifecycle
2017/05/25 22:48:15 Creating volume [basic_test_volume_1495752495] on VM [10.192.51.202]
2017/05/25 22:48:17 Checking volume [basic_test_volume_1495752495] availability from VM [10.192.51.202]
2017/05/25 22:48:17 Attaching volume [basic_test_volume_1495752495] on VM [10.192.51.202]
2017/05/25 22:48:19 Confirming attached status for volume [basic_test_volume_1495752495]
2017/05/25 22:48:21 Removing container [basic_test_container_1495752495] on VM [10.192.51.202]
2017/05/25 22:48:22 Confirming detached status for volume [basic_test_volume_1495752495]
2017/05/25 22:48:25 Destroying volume [basic_test_volume_1495752495]
2017/05/25 22:48:26 Checking volume [basic_test_volume_1495752495] availability from VM [10.192.51.202]
2017/05/25 22:48:26 Creating volume [basic_test_volume_1495752495] on VM [10.192.51.202]
2017/05/25 22:48:29 Checking volume [basic_test_volume_1495752495] availability from VM [10.192.51.202]
2017/05/25 22:48:29 Attaching volume [basic_test_volume_1495752495] on VM [10.192.51.202]
2017/05/25 22:48:31 Confirming attached status for volume [basic_test_volume_1495752495]
2017/05/25 22:48:33 Removing container [basic_test_container_1495752495] on VM [10.192.51.202]
2017/05/25 22:48:34 Confirming detached status for volume [basic_test_volume_1495752495]
2017/05/25 22:48:37 Destroying volume [basic_test_volume_1495752495]
2017/05/25 22:48:38 Checking volume [basic_test_volume_1495752495] availability from VM [10.192.51.202]
2017/05/25 22:48:38 END: basic-test.TestVolumeLifecycle
2017/05/25 22:48:38 Creating volume [restart_test_volume_1495752518] on VM [10.192.51.202]
2017/05/25 22:48:40 START: restart_test.TestPluginKill
2017/05/25 22:48:40 Attaching volume [restart_test_volume_1495752518] on VM[10.192.51.202]
2017/05/25 22:48:42 Confirming attached status for volume [restart_test_volume_1495752518]
2017/05/25 22:48:44 Killing vDVS plugin on VM [10.192.51.202]
2017/05/25 22:48:45 Sleep for 2 seconds
2017/05/25 22:48:47 Stopping container [restart_test_container_1495752518] on VM [10.192.51.202]
2017/05/25 22:48:59 Starting container [restart_test_container_1495752518] on VM [10.192.51.202]
2017/05/25 22:49:00 Stopping container [restart_test_container_1495752518] on VM [10.192.51.202]
2017/05/25 22:49:11 Confirming detached status for volume [restart_test_volume_1495752518]
2017/05/25 22:49:14 END: restart_test.TestPluginKill
2017/05/25 22:49:14 Removing container [restart_test_container_1495752518] on VM [10.192.51.202]
2017/05/25 22:49:14 Destroying volume [restart_test_volume_1495752518]
2017/05/25 22:49:15 Creating volume [restart_test_volume_1495752555] on VM [10.192.51.202]
2017/05/25 22:49:17 START: restart_test.TestVolumeStaleMount
2017/05/25 22:49:17 Attaching volume [restart_test_volume_1495752555] on VM [10.192.51.202]
2017/05/25 22:49:18 Confirming attached status for volume [restart_test_volume_1495752555]
2017/05/25 22:49:20 Killing docker on VM [10.192.51.202]
2017/05/25 22:49:21 Sleep for 2 seconds
2017/05/25 22:49:23 Confirming detached status for volume [restart_test_volume_1495752555]
2017/05/25 22:49:47 END: restart_test.TestVolumeStaleMount
2017/05/25 22:49:47 Removing container [restart_test_container_1495752555] on VM [10.192.51.202]
2017/05/25 22:49:47 Destroying volume [restart_test_volume_1495752555]
2017/05/25 22:49:48 Finding VM name from IP Address [10.192.51.202]
2017/05/25 22:49:49 START: Test vmdkops service restart + kill vm process
2017/05/25 22:49:49 Creating volume [vmlistener_test_volume_1495752589] on VM [10.192.51.202]
2017/05/25 22:49:51 Attaching volume [vmlistener_test_volume_1495752589] on VM [10.192.51.202]
2017/05/25 22:49:53 Confirming attached status for volume [vmlistener_test_volume_1495752589]
2017/05/25 22:49:54 Restarting vmdkops service on ESX host [10.192.59.225]
2017/05/25 22:49:58 Confirming attached status for volume [vmlistener_test_volume_1495752589]
2017/05/25 22:50:01 Retrieving VM power state for [ubuntu-VM0.2]
2017/05/25 22:50:02 VM[ubuntu-VM0.2]'s current power state is [poweredOn]
2017/05/25 22:50:02 Retrieving World ID for VM [ubuntu-VM0.2] from ESX [10.192.59.225]
2017/05/25 22:50:03 VM's process ID is: 2046140
2017/05/25 22:50:03 Killing VM from ESX [10.192.59.225]
2017/05/25 22:50:04 Sleep for 10 seconds
2017/05/25 22:50:14 Retrieving VM power state for [ubuntu-VM0.2]
2017/05/25 22:50:15 VM[ubuntu-VM0.2]'s current power state is [poweredOff]
2017/05/25 22:50:16 Powering on VM [ubuntu-VM0.2]
2017/05/25 22:50:18 Sleep for 35 seconds
2017/05/25 22:50:53 Retrieving VM power state for [ubuntu-VM0.2]
2017/05/25 22:50:54 VM[ubuntu-VM0.2]'s current power state is [poweredOn]
2017/05/25 22:50:54 Confirming detached status for volume [vmlistener_test_volume_1495752589]
2017/05/25 22:50:58 END: Test vmdkops service restart + kill vm process
2017/05/25 22:50:58 Destroying volume [vmlistener_test_volume_1495752589]
2017/05/25 22:50:58 Finding Datastores available on ESX
2017/05/25 22:50:59 Datastores found are [sharedVmfs-0 sharedVmfs-1 local.2-0 local.2-1 local.2-2 vsanDatastore]
2017/05/25 22:50:59 Creating volume [BpLnfgDsc2WD8F2qNfHK5a84jjJkwzDkh9h2fhfUVuS9jZ8uVbhV3vC5AWX39IVUWSP2NcHciWvqZTa2N95RxRTZHWUsaD6HEdz0T] on VM [10.192.51.202]
2017/05/25 22:51:00 Creating volume [Volume-000000] on VM [10.192.51.202]
2017/05/25 22:51:01 Creating volume [Volume_volume_1495752659@invalidDatastore] on VM [10.192.51.202]
2017/05/25 22:51:01 Creating volume [invalid_opts_volume_1495752661] with options [ -o diskformat=zeroedthickk] on VM [10.192.51.202]
2017/05/25 22:51:02 Creating volume [invalid_opts_volume_1495752662] with options [ -o diskformat=zeroedthick,thin] on VM [10.192.51.202]
2017/05/25 22:51:03 Creating volume [invalid_opts_volume_1495752663] with options [ -o size=100mbb] on VM [10.192.51.202]
2017/05/25 22:51:03 Creating volume [invalid_opts_volume_1495752663] with options [ -o size=100gbEE] on VM [10.192.51.202]
2017/05/25 22:51:04 Creating volume [invalid_opts_volume_1495752664] with options [ -o sizes=100mb] on VM [10.192.51.202]
2017/05/25 22:51:05 Creating volume [invalid_opts_volume_1495752665] with options [ -o fstype=xfs_ext] on VM [10.192.51.202]
2017/05/25 22:51:06 Creating volume [invalid_opts_volume_1495752666] with options [ -o access=read-write-both] on VM [10.192.51.202]
2017/05/25 22:51:06 Creating volume [invalid_opts_volume_1495752666] with options [ -o access=write-only] on VM [10.192.51.202]
2017/05/25 22:51:07 Creating volume [invalid_opts_volume_1495752667] with options [ -o access=read-write-both] on VM [10.192.51.202]
2017/05/25 22:51:07 Creating volume [invalid_opts_volume_1495752667] with options [ -o clone-from=IDontExist] on VM [10.192.51.202]
2017/05/25 22:51:08 Creating volume [hbXfQ6pYSQ3n267l1VQKGNbSuJE9fQbzONJAAwdCxmM8BIabKERsUhPNmMmdf2eSJyYtqwcFiUILzXv2fcNIrWO7sToFgoilA0U1] on VM [10.192.51.202]
2017/05/25 22:51:10 Checking volume [hbXfQ6pYSQ3n267l1VQKGNbSuJE9fQbzONJAAwdCxmM8BIabKERsUhPNmMmdf2eSJyYtqwcFiUILzXv2fcNIrWO7sToFgoilA0U1] availability from VM [10.192.51.202]
2017/05/25 22:51:11 Creating volume [Volume-0000000-****-###] on VM [10.192.51.202]
2017/05/25 22:51:13 Checking volume [Volume-0000000-****-###] availability from VM [10.192.51.202]
2017/05/25 22:51:14 Creating volume [Volume-00000] on VM [10.192.51.202]
2017/05/25 22:51:16 Checking volume [Volume-00000] availability from VM [10.192.51.202]
2017/05/25 22:51:17 Creating volume [Volume-0000000] on VM [10.192.51.202]
2017/05/25 22:51:19 Checking volume [Volume-0000000] availability from VM [10.192.51.202]
2017/05/25 22:51:20 Creating volume [abc_volume_1495752668@sharedVmfs-0] on VM [10.192.51.202]
2017/05/25 22:51:22 Checking volume [abc_volume_1495752668@sharedVmfs-0] availability from VM [10.192.51.202]
2017/05/25 22:51:22 Creating volume [abc_volume_1495752668@@@@sharedVmfs-0] on VM [10.192.51.202]
2017/05/25 22:51:24 Checking volume [abc_volume_1495752668@@@@sharedVmfs-0] availability from VM [10.192.51.202]
2017/05/25 22:51:25 Creating volume [Volume-你_volume_1495752668] on VM [10.192.51.202]
2017/05/25 22:51:27 Checking volume [Volume-你_volume_1495752668] availability from VM [10.192.51.202]
2017/05/25 22:51:28 Destroying volume [hbXfQ6pYSQ3n267l1VQKGNbSuJE9fQbzONJAAwdCxmM8BIabKERsUhPNmMmdf2eSJyYtqwcFiUILzXv2fcNIrWO7sToFgoilA0U1]
2017/05/25 22:51:29 Destroying volume [Volume-0000000-****-###]
2017/05/25 22:51:29 Destroying volume [Volume-00000]
2017/05/25 22:51:30 Destroying volume [Volume-0000000]
2017/05/25 22:51:31 Destroying volume [abc_volume_1495752668@sharedVmfs-0]
2017/05/25 22:51:32 Destroying volume [abc_volume_1495752668@@@@sharedVmfs-0]
2017/05/25 22:51:33 Destroying volume [Volume-你_volume_1495752668]
2017/05/25 22:51:34 Creating volume [clone_src_volume_1495752694] on VM [10.192.51.202]
2017/05/25 22:51:36 Creating volume [valid_opts_volume_1495752696] with options [ -o size=10gb] on VM [10.192.51.202]
2017/05/25 22:51:40 Checking volume [valid_opts_volume_1495752696] availability from VM [10.192.51.202]
2017/05/25 22:51:41 Creating volume [valid_opts_volume_1495752701] with options [ -o diskformat=zeroedthick] on VM [10.192.51.202]
2017/05/25 22:51:43 Checking volume [valid_opts_volume_1495752701] availability from VM [10.192.51.202]
2017/05/25 22:51:44 Creating volume [valid_opts_volume_1495752704] with options [ -o diskformat=thin] on VM [10.192.51.202]
2017/05/25 22:51:46 Checking volume [valid_opts_volume_1495752704] availability from VM [10.192.51.202]
2017/05/25 22:51:46 Creating volume [valid_opts_volume_1495752706] with options [ -o diskformat=eagerzeroedthick] on VM [10.192.51.202]
2017/05/25 22:51:49 Checking volume [valid_opts_volume_1495752706] availability from VM [10.192.51.202]
2017/05/25 22:51:49 Creating volume [valid_opts_volume_1495752709] with options [ -o attach-as=independent_persistent] on VM [10.192.51.202]
2017/05/25 22:51:51 Checking volume [valid_opts_volume_1495752709] availability from VM [10.192.51.202]
2017/05/25 22:51:52 Creating volume [valid_opts_volume_1495752712] with options [ -o attach-as=persistent] on VM [10.192.51.202]
2017/05/25 22:51:54 Checking volume [valid_opts_volume_1495752712] availability from VM [10.192.51.202]
2017/05/25 22:51:55 Creating volume [valid_opts_volume_1495752715] with options [ -o fstype=ext4] on VM [10.192.51.202]
2017/05/25 22:51:57 Checking volume [valid_opts_volume_1495752715] availability from VM [10.192.51.202]
2017/05/25 22:51:57 Creating volume [valid_opts_volume_1495752717] with options [ -o access=read-only] on VM [10.192.51.202]
2017/05/25 22:51:59 Checking volume [valid_opts_volume_1495752717] availability from VM [10.192.51.202]
2017/05/25 22:52:00 Creating volume [valid_opts_volume_1495752720] with options [ -o access=read-write] on VM [10.192.51.202]
2017/05/25 22:52:02 Checking volume [valid_opts_volume_1495752720] availability from VM [10.192.51.202]
2017/05/25 22:52:03 Creating volume [valid_opts_volume_1495752723] with options [ -o clone-from=clone_src_volume_1495752694] on VM [10.192.51.202]
2017/05/25 22:52:05 Checking volume [valid_opts_volume_1495752723] availability from VM [10.192.51.202]
2017/05/25 22:52:06 Creating volume [WxNeW1gdgUVD] with options [ -o fstype=xfs] on VM [10.192.51.202]
2017/05/25 22:52:08 Checking volume [WxNeW1gdgUVD] availability from VM [10.192.51.202]
2017/05/25 22:52:09 Destroying volume [clone_src_volume_1495752694]
2017/05/25 22:52:10 Destroying volume [valid_opts_volume_1495752696]
2017/05/25 22:52:10 Destroying volume [valid_opts_volume_1495752701]
2017/05/25 22:52:11 Destroying volume [valid_opts_volume_1495752704]
2017/05/25 22:52:12 Destroying volume [valid_opts_volume_1495752706]
2017/05/25 22:52:13 Destroying volume [valid_opts_volume_1495752709]
2017/05/25 22:52:14 Destroying volume [valid_opts_volume_1495752712]
2017/05/25 22:52:14 Destroying volume [valid_opts_volume_1495752715]
2017/05/25 22:52:15 Destroying volume [valid_opts_volume_1495752717]
2017/05/25 22:52:16 Destroying volume [valid_opts_volume_1495752720]
2017/05/25 22:52:17 Destroying volume [valid_opts_volume_1495752723]
2017/05/25 22:52:18 Destroying volume [WxNeW1gdgUVD]
2017/05/25 22:52:18 Finding Datastores available on ESX
2017/05/25 22:52:19 Creating policy [validPolicy] with content ['(("proportionalCapacity" i50)''("hostFailuresToTolerate" i0))'] on ESX [10.192.59.225]
2017/05/25 22:52:20 Creating policy [invalidPolicy] with content ['(("wrongKey" i50)'] on ESX [10.192.59.225]
2017/05/25 22:52:21 Creating volume [vsanVol_volume_1495752741@vsanDatastore] with options [ -o vsan-policy-name=validPolicy] on VM [10.192.51.202]
2017/05/25 22:52:24 Checking volume [vsanVol_volume_1495752741@vsanDatastore] availability from VM [10.192.51.202]
2017/05/25 22:52:25 Creating volume [vsanVol_volume_1495752745@vsanDatastore] with options [-o vsan-policy-name=IDontExist] on VM [10.192.51.202]
2017/05/25 22:52:25 Creating volume [vsanVol_volume_1495752745@vsanDatastore] with options [-o vsan-policy-name=invalidPolicy] on VM [10.192.51.202]
2017/05/25 22:52:27 Destroying volume [vsanVol_volume_1495752741@vsanDatastore]
OK: 10 passed
--- PASS: Test (257.52s)
```

shaominchen commented on tests/utils/inputparams/testparams.go:
```diff
@@ -53,6 +54,15 @@ func GetContainerNameWithTimeStamp(containerName string) string {
 }
 
+func GetVolumeNameOfSize(size int) string {
+	const chars = "abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ0123456789"
```
Nit: I'm fine with this special volume name. Just a little bit concern about the consistency: as we have discussed before, all test cases should create volumes (and other resources) with name starting with the test case name as the prefix, such as "restart_test_volume_xxxxx". The intention is to make it easier to keep track of volumes that are not cleaned up properly after running. I'm just thinking should we still follow that or not. Personally I tend to discard that convention, as it doesn't make much sense to me. What do you think? @pshahzeb @shuklanirdesh82

shuklanirdesh82 replied:
While eyeballing the test execution it is lot easier to know how far test execution has reached and what is going on now? on the other hand it is just a personal opinion, a lot more convenient to me; as we go along test name is our goto keyword like "refcount volume", "cloneVol" that helps to quickly know about the test it belongs. It is pretty handy to have referenced test name.

We may treat some special case as exception like L58.

We can use `C.TestName()` that gocheck is providing to get the current testcase name to keep it consistent.

pshahzeb replied:
I think we can go ahead with prefixes, but some test cases where we require exact sized volume names, we can make an exception to keep the util simple.

shaominchen replied:
Fine to me.

shuklanirdesh82 commented on tests/e2e/volumecreate_test.go:
```diff
@@ -0,0 +1,181 @@
+}
+
+func (s *VolumeCreateTestSuite) TearDownTest(c *C) {
+	for _, name := range s.volumeList {
```
I would suggest to create a volume name string with `space` separated and pass down to deleteVolume() function .. something like `docker volume rm vol1 vo2 ` that avoids subsequent ssh calls... get the response of the command and look for any `error` string . if not present we are good to go.

shuklanirdesh82 commented on tests/e2e/vsan_test.go:
```diff
@@ -0,0 +1,97 @@
+	dsNameList := govc.GetDatastoreList()
+	s.vsanDSName = ""
+	for _, ds := range dsNameList {
+		if strings.HasPrefix(ds, "vsan") {
```
this can work ... more robust and correct approach is to look for datastore type using govc.

pshahzeb commented on tests/e2e/volumecreate_test.go:
```diff
@@ -0,0 +1,181 @@
+}
+
+func (s *VolumeCreateTestSuite) SetUpTest(c *C) {
+	s.volumeList = s.volumeList[:0]
```
Sure

shaominchen commented on tests/e2e/volumecreate_test.go:
```diff
@@ -67,6 +72,7 @@ var _ = Suite(&VolumeCreateTestSuite{})
 // 6. contains multiple '@'
 // 7. contains unicode character
 func (s *VolumeCreateTestSuite) TestValidName(c *C) {
+	log.Printf("START: VolumeCreateTestSuite.TestValidName")
```
Nit: in existing tests we are using the file name (e.g. "START: restart_test.TestVolumeStaleMount"). I'm fine with both styles, just keep them consistent.

pshahzeb replied:
sure. can go ahead with file name.

shuklanirdesh82 commented on tests/utils/govc/cmd.go:
```diff
@@ -50,3 +51,16 @@ func PowerOnVM(vmName string) string {
 	return ssh.InvokeCommandLocally(cmd)
 }
+
+func GetDatastoreList() []string {
```
minor: need method doc

shuklanirdesh82 commented on tests/utils/govc/cmd.go:
```diff
@@ -50,3 +51,16 @@ func PowerOnVM(vmName string) string {
+	return strings.Fields(out)
+}
+
+func GetVsanDatastore() string {
```
same as above

shuklanirdesh82 commented on tests/utils/govc/cmd.go:
```diff
@@ -50,3 +51,16 @@ func PowerOnVM(vmName string) string {
+
+func GetVsanDatastore() string {
+	cmd := govc.DatastoreInfo + govc.JSONTypeOutput + "| " + govc.JSONParser + " '.Datastores[].Summary | select(.Type==\"vsan\") | .Name'"
+	out := ssh.InvokeCommandLocally(cmd)
```
we can return from this point.

pshahzeb commented:
@shaominchen @shuklanirdesh82 please take a look at my latest commit.
It addresses the nits indicated. It also adds a new test about volume access. I added it here for the dependency on the datastore util and to keep things simple.
Thanks.

shuklanirdesh82 commented on tests/e2e/volumecreate_test.go:
```diff
@@ -0,0 +1,201 @@
+	s.volumeList = append(s.volumeList, inputparams.GetVolumeNameWithTimeStamp("abc")+"@@@@"+s.dsNameList[0])
+	s.volumeList = append(s.volumeList, inputparams.GetVolumeNameWithTimeStamp("Volume-你"))
+
+	for _, name := range s.volumeList {
```
As discussed offline, It is better make them running in parallel instead of serially.

pshahzeb replied:
made the creation parallel.

shuklanirdesh82 commented on tests/e2e/volume_access_test.go:
```diff
@@ -0,0 +1,121 @@
+
+// The goal of this test suite is to verify read/write consistency on volumes
+// in accordance with the access updates on the volume
+package e2e
```
Let's create another target to run this test only once.

pshahzeb replied:
Can we do this in a separate quick PR?

shuklanirdesh82 replied:
yeah I will take care.

shaominchen commented on tests/e2e/volume_access_test.go:
```diff
@@ -0,0 +1,121 @@
+	s.dockerHostIP = []string{os.Getenv("VM1"), os.Getenv("VM2")}
+	s.esxIP = os.Getenv("ESX")
+
+	dsName := govc.GetDatastoreList()[0]
```
Are we expecting "datastore1" here? Please at least add one-line comment to clarify.

pshahzeb replied:
Sure. Will add the comment.. This can be any datastore. There should be atleast one (where the vm resides). The name can be datastore1 in case of CI, or anything else.

shuklanirdesh82 replied:
safer to check datastoreList is not empty before accessing its first element.

pshahzeb replied:
done

shuklanirdesh82 commented on tests/e2e/volume_access_test.go:
```diff
@@ -0,0 +1,121 @@
+	"github.com/vmware/docker-volume-vsphere/tests/utils/inputparams"
+)
+
+const ErrorWriteVolume = "Read-only file system"
```
do we really need to export this const?

shaominchen replied:
+1. Keep in mind that capital letter in Go means to be exported.

pshahzeb replied:
not exporting this const since its not used anywhere outside this test.

shuklanirdesh82 commented on tests/e2e/volume_access_test.go:
```diff
@@ -0,0 +1,121 @@
+func (s *VolumeAccessTestSuite) TearDownTest(c *C) {
+	for i := 0; i < 2; i++ {
+		for _, cname := range s.containerList[i] {
+			out, err := dockercli.RemoveContainer(s.dockerHostIP[i], cname)
```
I would suggest to remove containers into one shot .. e.g. `docker rm c1 c2`

govint replied:
This can be a common tear down function that all tests can use, its not specific to this test. Since all containers are named in a certain way they can be identified as "docker ps -aq | grep test container name pattern" and remove those.

pshahzeb replied:
using join and removing them in one shot.

shuklanirdesh82 commented on tests/e2e/volume_access_test.go:
```diff
@@ -0,0 +1,121 @@
+	c.Assert(out, Equals, data1)
+
+	out, err = dockercli.WriteToVolume(s.dockerHostIP[1], s.volumeName, s.newCName(1), testFile, data2)
+	c.Assert(err, IsNil, Commentf(out))
```
don't we need to verify written content for the L#99?

pshahzeb replied:
added extra check

shuklanirdesh82 commented on tests/e2e/volume_access_test.go:
```diff
@@ -0,0 +1,121 @@
+	out, err = dockercli.WriteToVolume(s.dockerHostIP[1], s.volumeName, s.newCName(1), testFile, data2)
+	c.Assert(err, IsNil, Commentf(out))
+
+	out, err = admincli.UpdateVolumeAccess(s.esxIP, s.volumeName, "_DEFAULT", "read-only")
```
`"_DEFAULT"` ... I believe there is a const from vmgroup/admincli package, need to use it here.

pshahzeb replied:
done

shuklanirdesh82 commented on tests/e2e/volume_access_test.go:
```diff
@@ -0,0 +1,121 @@
+	out, err = dockercli.WriteToVolume(s.dockerHostIP[1], s.volumeName, s.newCName(1), testFile, data2)
+	c.Assert(strings.Contains(out, ErrorWriteVolume), Equals, true, Commentf(out))
+
+	out, err = admincli.UpdateVolumeAccess(s.esxIP, s.volumeName, "_DEFAULT", "read-write")
```
same as above.

pshahzeb replied:
fixed

shuklanirdesh82 commented on tests/e2e/volume_access_test.go:
```diff
@@ -0,0 +1,121 @@
+
+	out, err = dockercli.WriteToVolume(s.dockerHostIP[0], s.volumeName, s.newCName(0), testFile, data1)
+	c.Assert(err, IsNil, Commentf(out))
+
```
need to verify content here to make sure volumes are writable now.

pshahzeb replied:
fixed

shuklanirdesh82 commented on tests/e2e/volumecreate_test.go:
```diff
@@ -0,0 +1,201 @@
+	. "gopkg.in/check.v1"
+)
+
+const ErrorVolumeCreate = "Error response from daemon: create"
```
1. `ErrorVolumeCreate ` => `volumeCreationError` 
2. do we really need to export this const variable?

pshahzeb replied:
Since this is an error constant, i wanted the name to begin with error.
This constant is also used in vsan test where create volume fails.
Not sure if it is worth adding to a separate error constants and create a separate package for it.

shuklanirdesh82 commented on tests/e2e/volumecreate_test.go:
```diff
@@ -0,0 +1,201 @@
+
+func (s *VolumeCreateTestSuite) SetUpTest(c *C) {
+	// clean the list of volumes created
+	s.volumeList = s.volumeList[:0]
```
shouldn't this be part of tearDown?

pshahzeb replied:
fixed

shuklanirdesh82 commented on tests/e2e/volumecreate_test.go:
```diff
@@ -0,0 +1,201 @@
+
+func (s *VolumeCreateTestSuite) TearDownTest(c *C) {
+	volList := ""
+	for _, name := range s.volumeList {
```
Ideally no need to have this extra for loop here .. it should be part of test method itself. If volume is created then add to the cleanup list in this case `s.volumeList`.

pshahzeb replied:
removed. using join instead.

shuklanirdesh82 commented on tests/e2e/volumecreate_test.go:
```diff
@@ -0,0 +1,201 @@
+	invalidVolList = append(invalidVolList, inputparams.GetVolumeNameWithTimeStamp("Volume")+"@invalidDatastore")
+
+	for _, name := range invalidVolList {
+		out, _ := dockercli.CreateVolume(s.hostIP, name)
```
is there a reason for ignoring error here?

pshahzeb replied:
Refactored this code. err is not nil and output message contains create error substring

shuklanirdesh82 commented on tests/e2e/volumecreate_test.go:
```diff
@@ -0,0 +1,201 @@
+
+	for _, name := range invalidVolList {
+		out, _ := dockercli.CreateVolume(s.hostIP, name)
+		c.Assert(strings.HasPrefix(out, ErrorVolumeCreate), Equals, true)
```
this works fine in happy path/scenario .. how are we handling cleaning up for crated volume when there is a bug in production code and it creates the volume with invalid input?

pshahzeb replied:
refactored the create code. Adding volume name to list in case it is created so that it gets cleaned appropriately.

shuklanirdesh82 commented on tests/e2e/vsan_test.go:
```diff
@@ -0,0 +1,95 @@
+	c.Assert(err, IsNil, Commentf(out))
+
+	s.volumeName = inputparams.GetVolumeNameWithTimeStamp("vsanVol") + "@" + s.vsanDSName
+	vsanOpts := " -o vsan-policy-name=" + policyName
```
`vsan-policy-name` should be moved as constant ... it will be easier to make a change when the parameter name got changes .. earlier we were referring it as `policy-name` and it is likely to be change as `vsan` is too specific.

pshahzeb replied:
done

shuklanirdesh82 commented on tests/utils/dockercli/volumelifecycle.go:
```diff
@@ -51,6 +57,26 @@ func AttachVolumeWithRestart(ip, volName, containerName string) (string, error)
 
+// WriteToVolume write data to a given file on given volume
+func WriteToVolume(ip, volName, containerName, fileName, data string) (string, error) {
+	log.Printf("Writing %s to file %s on volume [%s] on VM[%s]\n", data, fileName, volName, ip)
```
nit: `on VM` => `from/through VM`

pshahzeb replied:
done in #1320

govint commented on tests/utils/dockercli/volumelifecycle.go:
```diff
@@ -51,6 +57,26 @@ func AttachVolumeWithRestart(ip, volName, containerName string) (string, error)
+	log.Printf("Reading from file %s on volume [%s] on VM[%s]\n", fileName, volName, ip)
+
+	readCmd := " /bin/sh -c 'cat /vol1/" + fileName + "'"
+	return ssh.InvokeCommand(ip, dockercli.RunContainer+" -v "+volName+
```
dockercli.RunContainer doesn;t include the "-rm" flag as it should, so, this container will remain in the system and won;t be removed in the teardown function. Please update dockercli.RunContainer to include "-rm".

pshahzeb replied:
in teardown we call remove container which essentially does docker rm.
In some test cases we need to stop the container and start the same container again. So not adding -rm flag in RunContainer because the container would be removed after it exits if I add the flag.

govint commented on tests/utils/govc/cmd.go:
```diff
@@ -50,3 +51,17 @@ func PowerOnVM(vmName string) string {
+func GetVsanDatastore() string {
+	cmd := govc.DatastoreInfo + govc.JSONTypeOutput + "| " + govc.JSONParser + " '.Datastores[].Summary | select(.Type==\"vsan\") | .Name'"
+	return ssh.InvokeCommandLocally(cmd)
+}
```
Please include this in the PR for just the common code and not part of this test.

Why do we need only a get() for VSAN DS

pshahzeb replied:
Need the vsan datastore name for vsan test suite - to create volume on vsandatastore.
Making this util more generic to get datastore of given datatype. 
#1320

pshahzeb commented on tests/constants/dockercli/cmd.go:
```diff
@@ -71,8 +71,11 @@ const (
+
 	// TestContainer test busybox container that keeps running
-	TestContainer = " busybox tail -f /dev/null "
+	TestContainer = ContainerImage + " tail -f /dev/null "
```
Yes we can have a common teardown function. And if need be to have another container with given sleep time, we can add as required. Keeping this as it is but moving to separate PR #1320

pshahzeb commented on tests/constants/govc/cli.go:
```diff
@@ -25,6 +25,12 @@ const (
+
+	// DatastoreList get the list of datastore names from govc output
+	DatastoreList = ".Datastores[].Name "
+
```
#1320

pshahzeb commented on tests/utils/dockercli/volumelifecycle.go:
```diff
@@ -33,6 +33,12 @@ func CreateVolume(ip, name string) (string, error) {
 }
 
+// CreateVolumeWithOptions is going to create vsphere docker volume with given name.
+func CreateVolumeWithOptions(ip, name, options string) (string, error) {
```
#1320

shuklanirdesh82 commented:
This PR is not just for volume create infact it has plugin tests as well. 

- Need to update PR title.
- Need to add `fixes` (issue number)

pshahzeb commented on tests/e2e/volumecreate_test.go:
```diff
@@ -0,0 +1,201 @@
+
+	var validVolOpts []string
+
+	validVolOpts = append(validVolOpts, " -o size=10gb")
```
Done

pshahzeb commented on tests/e2e/volumecreate_test.go:
```diff
@@ -0,0 +1,201 @@
+
+	var invalidVolList []string
+
+	invalidVolList = append(invalidVolList, inputparams.GetVolumeNameOfSize(101))
```
Done

pshahzeb commented on tests/e2e/volumecreate_test.go:
```diff
@@ -0,0 +1,201 @@
+	s.volumeList = append(s.volumeList, "Volume-0000000")
+	s.volumeList = append(s.volumeList, inputparams.GetVolumeNameWithTimeStamp("abc")+"@"+s.dsNameList[0])
+	s.volumeList = append(s.volumeList, inputparams.GetVolumeNameWithTimeStamp("abc")+"@@@@"+s.dsNameList[0])
+	s.volumeList = append(s.volumeList, inputparams.GetVolumeNameWithTimeStamp("Volume-你"))
```
Done.

pshahzeb commented on tests/e2e/volumecreate_test.go:
```diff
@@ -0,0 +1,201 @@
+	log.Printf("START: volumecreate_test.TestValidName")
+
+	s.volumeList = append(s.volumeList, inputparams.GetVolumeNameOfSize(100))
+	s.volumeList = append(s.volumeList, "Volume-0000000-****-###")
```
Using the string array, but not making it global because it will be used only inside this test func

pshahzeb commented on tests/e2e/volumecreate_test.go:
```diff
@@ -0,0 +1,201 @@
+
+// Valid volume names test
+// 1. having 100 chars
+// 2. having various chars including alphanumerics
```
Added extra valid test case for spaced name.

pshahzeb commented on tests/e2e/volumecreate_test.go:
```diff
@@ -0,0 +1,201 @@
+func (s *VolumeCreateTestSuite) TearDownTest(c *C) {
+	volList := ""
+	for _, name := range s.volumeList {
+		volList = volList + " " + name
```
fixed

pshahzeb commented on tests/e2e/volume_access_test.go:
```diff
@@ -0,0 +1,121 @@
+	c.Assert(strings.Contains(out, ErrorWriteVolume), Equals, true, Commentf(out))
+
+	out, err = dockercli.WriteToVolume(s.dockerHostIP[1], s.volumeName, s.newCName(1), testFile, data2)
+	c.Assert(strings.Contains(out, ErrorWriteVolume), Equals, true, Commentf(out))
```
I wanted to capture the substring of failure to write on the volume. err is not nil but doesn't help much to identify write error.

pshahzeb commented on tests/e2e/volume_access_test.go:
```diff
@@ -0,0 +1,121 @@
+	out, err = admincli.UpdateVolumeAccess(s.esxIP, s.volumeName, "_DEFAULT", "read-only")
+	c.Assert(err, IsNil, Commentf(out))
+
+	out, err = dockercli.WriteToVolume(s.dockerHostIP[0], s.volumeName, s.newCName(0), testFile, data1)
```
Yes. Not sure if we need to add a detached status check here though.

pshahzeb commented on tests/utils/inputparams/testparams.go:
```diff
@@ -49,6 +49,11 @@ func GetVolumeNameWithTimeStamp(volumeName string) string {
 }
 
+// GetUniqueVolumeName prepares unique volume name with a random generated number
+func GetUniqueVolumeName(volName string) string {
```
Borrowed this from #1286  since i too ran into the issue while creating volumes in parallel. will remove this from here if it gets merged before

pshahzeb commented on tests/e2e/volumecreate_test.go:
```diff
@@ -0,0 +1,201 @@
+	esxIP      string
+	dsNameList []string
+	volumeList []string
+}
```
Sure. we can have a separate PR to have this common struct and remove it from the tests.

pshahzeb commented:
@shuklanirdesh82 
The updated steps for volume access steps in #820 still look overlapping to me.
Scenario 1 and 2 won't test any different code paths.

//CC @lipingxue

pshahzeb commented:
CI has passed
@shuklanirdesh82 I have made volume creation parallel and addressed your comments. 
@govint I have addressed your comments as well.

shuklanirdesh82 commented:
> https://github.com/vmware/docker-volume-vsphere/pull/1299#issuecomment-306252947

@pshahzeb as discussed offline we need to add that coverage and I will review later one with the change.

pshahzeb commented:
@shuklanirdesh82 As discussed, i have added another test case to access test to go from read-only volume to read-write.

shuklanirdesh82 commented on tests/e2e/volume_access_test.go:
```diff
@@ -0,0 +1,190 @@
+// The goal of this test suite is to verify read/write consistency on volumes
+// in accordance with the access updates on the volume
+
+// +build runalways
```
as discussed offline .. let's use `runonce`

shuklanirdesh82 commented on tests/utils/inputparams/testparams.go:
```diff
@@ -23,15 +23,16 @@ import (
 	"os"
 	"strconv"
 	"time"
+
```
need to remove this file from the PR .. seeing only formatting change.
